pkg/driverbuilder/builder/flatcar.go

```diff
@@ -0,0 +1,232 @@
+	}
+
+	gccVersion := ""
+	kernelVersion := ""
+	// structure of a package line is: category/name-version(-revision)::repository
+	for _, pkg := range strings.Split(string(packageList), "\n") {
+		if strings.HasPrefix(pkg, "sys-devel/gcc") {
+			gccVersion = pkg[len("sys-devel/gcc-"):]
+			gccVersion = strings.Split(gccVersion, "::")[0]
+			gccVersion = strings.Split(gccVersion, "-")[0]
+		}
+		if strings.HasPrefix(pkg, "sys-kernel/coreos-kernel") {
+			kernelVersion = pkg[len("sys-kernel/coreos-kernel-"):]
+			kernelVersion = strings.Split(kernelVersion, "::")[0]
+			kernelVersion = strings.Split(kernelVersion, "-")[0]
+		}
+	}
+	flatcarInfo.GCCVersion = gccVersion
+	flatcarInfo.KernelVersion = kernelVersion
+
+	return &flatcarInfo, nil
+}
+
+func fetchFlatcarPackageListURL(flatcarVersion string) []string {
+	pattern := "https://%s.release.flatcar-linux.net/amd64-usr/%s/flatcar_production_image_packages.txt"
+	channels := []string{
+		"stable",
+		"beta",
+		"alpha",
+	}
+	urls := []string{}
+	for _, channel := range channels {
+		urls = append(urls, fmt.Sprintf(pattern, channel, flatcarVersion))
+	}
+	return urls
+}
+
+func fetchFlatcarKernelConfigURL(flatcarChannel, flatcarVersion string) []string {
+	return []string{fmt.Sprintf("https://%s.release.flatcar-linux.net/amd64-usr/%s/flatcar_production_image_kernel_config.txt", flatcarChannel, flatcarVersion)}
+}
+
+func fetchFlatcarKernelURLS(kernelVersion string) []string {
+	kv := kernelrelease.FromString(kernelVersion)
+	return []string{fmt.Sprintf("https://cdn.kernel.org/pub/linux/kernel/v%s.x/linux-%s.tar.xz", kv.Version, kv.Fullversion)}
+}
+
+type flatcarReleaseInfo struct {
+	Channel       string
+	GCCVersion    string
+	KernelVersion string
+}
+
+type flatcarTemplateData struct {
+	DriverBuildDir    string
+	ModuleDownloadURL string
+	KernelDownloadURL string
+	GCCVersion        string
+	FlatcarVersion    string
+	FlatcarChannel    string
+	KernelConfigURL   string
+	ModuleDriverName  string
+	ModuleFullPath    string
+	BuildModule       bool
+	BuildProbe        bool
+}
+
+const flatcarTemplate = `
+#!/bin/bash
+set -xeuo pipefail
+
+rm -Rf {{ .DriverBuildDir }}
+mkdir {{ .DriverBuildDir }}
+rm -Rf /tmp/module-download
+mkdir -p /tmp/module-download
+
+curl --silent -SL {{ .ModuleDownloadURL }} | tar -xzf - -C /tmp/module-download
+mv /tmp/module-download/*/driver/* {{ .DriverBuildDir }}
+
+cp /driverkit/module-Makefile {{ .DriverBuildDir }}/Makefile
+bash /driverkit/fill-driver-config.sh {{ .DriverBuildDir }}
+
+# Fetch the kernel
+mkdir /tmp/kernel-download
+cd /tmp/kernel-download
+curl --silent -SL {{ .KernelDownloadURL }} | tar -Jxf - -C /tmp/kernel-download
+rm -Rf /tmp/kernel
+mkdir -p /tmp/kernel
+mv /tmp/kernel-download/*/* /tmp/kernel
+
+# Change current gcc
+ln -sf /usr/bin/gcc-{{ .GCCVersion }} /usr/bin/gcc
+
+curl --silent -o /tmp/kernel.config -SL {{ .KernelConfigURL }}
+
+cd /tmp/kernel
+make KCONFIG_CONFIG=/tmp/kernel.config oldconfig
```

Comment:
I wonder if these calls to make `oldconfig`, `modules_prepare` are necessary? I don't actually see them in any of the other builders, but this may be required for flatcar?

Reply:
I copied this part from the vanilla builder. Our own "kernel-devel" package equivalent is compiled against newer glibc, such that it doesn't work with the old debian based builder image. Here we're pulling generic kernel sources (we don't patch our sources) and need to "prepare" those.